### ISSUE TITLE
Canonical devcontainer feature with an up-to-date version of `fish`

### ISSUE THREAD

**Issue by lgarron** [fish-shell/fish-shell#9842]
GitHub Codespaces has the concept of "devcontainer features", which are sort of like package installation but with more batteries included: https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/adding-a-dev-container-configuration/introduction-to-dev-containers#adding-additional-features-to-your-devcontainerjson-file

https://containers.dev/features lists two published devcontainer features for `fish`:

- [`ghcr.io/devcontainers-contrib/features/fish-apt-get:1`](https://ghcr.io/devcontainers-contrib/features/fish-apt-get:1)
  - This one is super out of date — it installs `fish` 3.1. This is incompatible with the kinds of things you might put in `config.fish` for newer versions of `fish`, like the new abbreviation-related functionality in `fish` 3.6. The incompatibility tends to be… not graceful.
- [`ghcr.io/meaningful-ooo/devcontainer-features/fish:1`](https://ghcr.io/meaningful-ooo/devcontainer-features/fish:1)
  - This is someone's personal published devcontainer feature. I'd be glad to use it for my own projects, but this is a bit of a liability to use inside a company.
    - One mitigation might be to use a hash instead of a version number for auditable trust. But the devcontainer feature docs contain no info about pinning to a hash, nor do they provide any other assurances about immutability[^1]. 🤷

[^1]: Unless "Tooling that handles releasing Features will not republish Features if that exact version has already been published" https://containers.dev/implementors/features/ counts? That's rather weak.

I wanted to ask the `fish` maintainers: Is this something that you think should be maintained? If so, do you have a preference among the following?

- Add a new feature to `devcontainers-contrib` that installs the latest `fish`.
- Add a new repo to https://github.com/fish-shell to host the feature.
- Leave things as-is?

**Comment by faho** [comment]
For me personally I can say that I have no interest in maintaining this because I don't use Github Codespaces.

If someone else wants to that's fine and I'm not going to stop them.

>This one is super out of date — it installs fish 3.1

It installs fish *from the distro repos*, so it will pick whatever that ships. Fish 3.1 is what you get in Debian 11 or Ubuntu 20.04.

> Add a new feature to devcontainers-contrib that installs the latest fish.

This involves changing https://github.com/devcontainers-contrib/features/tree/main/src/fish-apt-get so it adds our repositories - either [OBS](https://software.opensuse.org/download.html?project=shells%3Afish%3Arelease%3A3&package=fish) or [launchpad](https://launchpad.net/~fish-shell/+archive/ubuntu/release-3).

This isn't *too* involved and can be done by someone who isn't a fish developer.

Note that the script currently supports alpine, which we don't have a repo for.

----

I'm going to leave this open for a week or so, if someone volunteers that's cool, if not I guess the default answer is 3: "Leave things as-is".

I'll tag it "needs-more-info" to remind me (I go through those periodically and close them if nothing happens).

**Comment by lgarron** [comment]
> This involves changing https://github.com/devcontainers-contrib/features/tree/main/src/fish-apt-get so it adds our repositories - either [OBS](https://software.opensuse.org/download.html?project=shells%3Afish%3Arelease%3A3&package=fish) or [launchpad](https://launchpad.net/~fish-shell/+archive/ubuntu/release-3).
> 
> This isn't _too_ involved and can be done by someone who isn't a fish developer.

I'd be happy to submit that! https://launchpad.net/~fish-shell/+archive/ubuntu/release-3 says this pretty much just involves:

```shell
sudo apt-add-repository ppa:fish-shell/release-3
sudo apt update
sudo apt install fish
```

Is there anything I should worry about, apart from a pared-down version of those calls?

**Comment by faho** [comment]
>I'd be happy to submit that! https://launchpad.net/~fish-shell/+archive/ubuntu/release-3 says this pretty much just involves:

That's the Ubuntu version, yeah. I'm not sure if apt-add-repository is still shipped by default (#9319 says it's deprecated), especially on containers.

If it isn't, you'd need to add the repo manually - launchpad says it's

```
deb https://ppa.launchpadcontent.net/fish-shell/release-3/ubuntu YOUR_UBUNTU_VERSION_HERE main 
```

where you can get the Ubuntu version from /etc/os-release - it has an "UBUNTU_CODENAME=codename" line.

You'll also have to import the key from https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x59fda1ce1b84b3fad89366c027557f056dc33ca5 and import that.

For other distros see OBS - it has instructions that you could probably just add directly (after checking /etc/os-release to determine the distro).

**Comment by faho** [comment]
I'm gonna close this on our end because it seems like you want to try getting it upstream first.

If there's any questions, do of course feel free to ask.
